src/math/wasm/truncf.c

```diff
@@ -0,0 +1,6 @@
+#include <math.h>
+
+__attribute__((const)) float truncf(float x)
+{
+	return __builtin_truncf(x);
+}
```

Comment:
You can also optimize `nearbyint`/`nearbyintf` and `rint`/`rintf` using `__builtin_nearbyint`/`__builtin_nearbyintf` and `__builtin_rint`/`__builtin_rintf`, respectively, which codegen to `f64.nearest`/`f32.nearest`. (This is possible for rint because wasm doesn't have user-accessible floating-point exceptions, and for nearbyint and rint because wasm doesn't have user-modifiable rounding modes.

Reply:
Thanks, added those too.

I did a bit of trial-and-error to work out which math builtins could be optimised this way. Adding `__attribute__((const))` was a bit of trial-and-error. Musl doesn't set errno for any maths functions, so it should be safe/consistent.

I was hoping to turn fmin/fmax into `f64.min`, but couldn't seem to get it to be emitted by `__builtin_fmax`. Those builtins are tricky, because if they don't lower to the expected native opcode, then they generate a recursive call to the libc function!

In any case, these routines in libc are pretty unimportant. LLVM replaces calls to these libc functions with the native opcodes anyway, so they won't/shouldn't actually be called here unless you're calling them by pointer.

Reply:
Can you think of a reason why this body doesn't emit a nice `f64.min` instruction:

`__attribute__((const)) double fmin(double x, double y) { return __builtin_fmin(x,y); }`

I'm not confident in saying it's a bug in the WebAssembly frontend, or whether my expectation is wrong, but it is surprising that so many of the other maths builtins do work.

Reply:
When one operand is NaN, wasm's `min` and `max` return NaN, while C's `fmin` and `fmax` return the other operand. Both forms are useful, so I expect that in the future wasm will have opcodes for both forms. As a historical note, the reason why wasm initially left out the latter form because IEEE 754-2008's definition of this form had surprising behavior on signaling NaN (see [here](https://2pi.dk/2016/05/ieee-min-max) for a description). The IEEE 754 committee has recently recognized this as an error, and drafts of the upcoming 754-2018 have removed that definition and provided a new corrected one.

Reply:
I see, thanks!